### PR TITLE
Keep the cursor position in the Classic Block

### LINKED ISSUE #10509: Classic Block - Insert Media Icon - New Upload jumps to top
When you insert a media file, e. g. an image with the insert media icon
then hit return 
and insert a second image the same way, the new image jumps to top, But it should be below the first image.

mediajump

**Desktop (please complete the following information):**
 - OS: Windows 10 (recent build) 64bit
 - Browser Chrome (recent build)
 - Version Gutenberg pre-release download package

### PR DESCRIPTION
**Pull request by azaozz**
Fixes #10509.

- On blur store the cursor position and the "raw" content of the Classic Block.
- Use the "raw" content when refreshing the editor content on `componentDidUpdate()`. It contains the cursor position "bookmark".
- Restore the cursor position on focus and before inserting content.

### CODE REVIEW

File: packages/block-library/src/classic/edit.js [hunk 1]
```diff
@@ -50,12 +50,20 @@ export default class ClassicEdit extends Component {
 	}
 
 	componentDidUpdate( prevProps ) {
-		const { clientId, attributes: { content } } = this.props;
-
+		const { clientId, attributes } = this.props;
 		const editor = window.tinymce.get( `editor-${ clientId }` );
+		let { contentRaw } = attributes;
```
Comment (youknowriad) [inline packages/block-library/src/classic/edit.js]: contentRaw is not persisted, we should probably use "local state" for it.

**Review by youknowriad — COMMENTED** [review]
We could use some e2e test for the classic block :)

File: packages/block-library/src/classic/edit.js (hunk 1 above)
Comment (azaozz) [reply packages/block-library/src/classic/edit.js]: Ah, sure, that'd work even better :)

File: packages/block-library/src/classic/edit.js [hunk 2]
```diff
@@ -84,12 +92,29 @@ export default class ClassicEdit extends Component {
 		}
 
 		editor.on( 'blur', () => {
+			editor.wpClassicBlockBookmark = editor.selection.getBookmark();
+
+			// Keep the "raw" content as it contains the caret position bookmark.
 			setAttributes( {
+				contentRaw: editor.getContent( { format: 'raw' } ),
```
Comment (iseulde) [inline packages/block-library/src/classic/edit.js]: Storing the raw content without any processing seems like a bad idea? Why should this be different from the classic editor?

File: packages/block-library/src/classic/edit.js [hunk 3]
```diff
@@ -84,12 +92,29 @@ export default class ClassicEdit extends Component {
...
 				content: editor.getContent(),
 			} );
 			return false;
 		} );
 
+		editor.on( 'init', () => {
+			// Store the initial "raw" content.
+			setAttributes( {
+				contentRaw: editor.getContent( { format: 'raw' } ),
```
Comment (iseulde) [inline packages/block-library/src/classic/edit.js]: I wouldn't expect merely initialising the block to change the content. The user didn't do  anything.

File: packages/block-library/src/classic/edit.js [hunk 4]
```diff
@@ -50,12 +50,20 @@ export default class ClassicEdit extends Component {
 	}
 
 	componentDidUpdate( prevProps ) {
-		const { clientId, attributes: { content } } = this.props;
-
+		const { clientId, attributes } = this.props;
 		const editor = window.tinymce.get( `editor-${ clientId }` );
+		let { contentRaw } = attributes;
+
+		// Strip white space before comparing to avoid resetting the editor content when not necessary;
```
Comment (iseulde) [inline packages/block-library/src/classic/edit.js]: What kind of whitespace is different?

**Comment by azaozz** [comment]
Looking more at this, still can't figure out what the refresh code in `componentDidUpdate()` is supposed to do:
```
const { clientId, attributes: { content } } = this.props;
....
if ( prevProps.attributes.content !== content ) {
	editor.setContent( content || '' );
}
```

Seems at some point it was possible that `attributes.contend` was changed/updated outside the editor, but that doesn't seem to be the case now. Perhaps this is a "left over" from how undo/redo was working previously?

Even if this is still needed, don't think that looking at `prevProps.attributes.content` is a good way to determine what's in the editor at that moment. It needs to do `editor.getContent()`.

@iseulde @aduth any chance you could remember what was the purpose of the above code? I know, it's been there for over a year...

**Comment by iseulde** [comment]
To me, this is an interesting thing that happens in master:

1. Load a classic block with some text. Place the caret somewhere other than the start.
2. Click the media button in the toolbar.
3. Select an image.
4. Close the media modal. Do not insert it!
5. Place the caret back where it was (somewhere other than the very start).
6. Click the media button in the toolbar.
7. The image you selected last time will already be selected. Do not click it again!
8. Insert the already selected image.
9. The image is correctly inserted.

This means that clicking the image specifically is causing the focus loss?

File: packages/block-library/src/classic/edit.js (hunk 4 above)
Comment (azaozz) [reply packages/block-library/src/classic/edit.js]: The only differences I'm seeing are in some `<p>\u00a0</p>` and some `;` missing/being added to ends of inline style attributes.

**Comment by iseulde** [comment]
> Seems at some point it was possible that attributes.contend was changed/updated outside the editor, but that doesn't seem to be the case now.

I think it is still needed for undo/redo. It needs to be possible to set the content from outside the classic block.

> Even if this is still needed, don't think that looking at prevProps.attributes.content is a good way to determine what's in the editor at that moment. It needs to do editor.getContent().

Sounds good to me.

**Comment by azaozz** [comment]
> This means that clicking the image specifically is causing the focus loss?

Yes, clicking anywhere outside the editor triggers the `blur()` and looses the caret position in the editor body <div>.

**Comment by iseulde** [comment]
Another interesting thing: even if I remove `componentDidUpdate` entirely, the problem still exists. Again only happens if you select the image in the media library. Otherwise note.

**Comment by azaozz** [comment]
> even if I remove componentDidUpdate entirely, the problem still exists

Right. It still needs `getBookmark()`, `moveToBookmark()` on blur/focus.

**Comment by iseulde** [comment]
> Right. It still needs getBookmark(), moveToBookmark() on blur/focus.

Why does this only happen for selecting an image though? What's special about this image?

**Comment by azaozz** [comment]
Here it happens for any click that focuses something outside the editor body. For example opening the menu, etc.

**Comment by azaozz** [comment]
Found it :) 

The check in componentDidUpdate() is needed only when the global undo/redo buttons are clicked. Not needed/not running for Ctrl/Cmd + z though...

**Comment by azaozz** [comment]
Closing this in favor of #12415.

**Pull request closed without merging**